powerline/segments/common/bat.py

```diff
@@ -136,41 +141,19 @@ class PowerClass(Structure):
 				('BatteryFullLifeTime', c_ulong)
 			]
 
-		def _get_capacity(pl):
-			powerclass = PowerClass()
-			result = library_loader.kernel32.GetSystemPowerStatus(byref(powerclass))
-			# http://msdn.microsoft.com/en-us/library/windows/desktop/aa372693(v=vs.85).aspx
-			if result:
-				return None
-			return powerclass.BatteryLifePercent
-
-		if _get_capacity() is None:
+		powerclass = PowerClass()
+		result = library_loader.kernel32.GetSystemPowerStatus(byref(powerclass))
+		# http://msdn.microsoft.com/en-us/library/windows/desktop/aa372693(v=vs.85).aspx
+		if result:
 			pl.debug('Not using GetSystemPowerStatus because it failed')
 		else:
 			pl.debug('Using GetSystemPowerStatus')
-
-		return _get_capacity
+			return powerclass.BatteryLifePercent, powerclass.ACLineStatus == 1
 
 	raise NotImplementedError
 
 
-def _get_capacity(pl):
-	global _get_capacity
-
-	def _failing_get_capacity(pl):
-		raise NotImplementedError
-
-	try:
-		_get_capacity = _get_battery(pl)
-	except NotImplementedError:
-		_get_capacity = _failing_get_capacity
-	except Exception as e:
-		pl.exception('Exception while obtaining battery capacity getter: {0}', str(e))
-		_get_capacity = _failing_get_capacity
```

Comment:
This is caching. `_get_battery` is called only once and it returns some function or fails with some exception. After this `_get_capacity` function is replaced either with returned function or with `_failing_get_capacity` and bold `_get_battery` function is *never called after this*.

Caching is saving the result of calculations in order to not repeat it. Calculations are determining which method should be used for obtaining battery capacity.